AirbyteLib SQLCaches for DuckDB and Postgres (also includes Ruff+Mypy cleanup)

aaronsteers opened the pull request:
Should merge after:

- #33409 

The latest on my PR does a bunch of things:
1. DuckDB SQL cache: working in basic integration test.
2. Postgres SQL cache: also working - via docker in pytest.
   - While we don't really need more than one SQL cache at this point, it only took about 3 hours to build the Postgres version (most of the time setting up the test harness), and it helped me validate some design patterns.
3. Automatic type conversions, from JSON Schema types and Airbyte types appears to be working - at least for the very simple tests. I didn't need to write any different code between Postgres and DuckDB, but some custom code might be needed.
4. Parquet file writing. The write to Parquet files is automatic and mostly transparent to the SQLCache classes.
5. The whole project (everything under airbyte_lib) now lints and formats with Ruff and passed mypy tests. I've added Ruff, isort, and mypy rules to pyproject.toml. (Also: isort is managed by Ruff now.)
6. I changed the integration tests to use a common fixture for the test data streams, and I compared the datasets using pandas (pandas.testing.assert_frame_equal()), which reduces complexity of the test code, and failures are easier to interpret now.
7. I added "FIXME": and "TODO:" notes in the code, so they are easy to flag for follow-up and aren't misinterpreted as random bugs. One example is 'in-memory' DuckDB databases; If we aren't going to need in-memory sql dbs, then a lot of that code can be simplified/removed.

aaronsteers commented:
Current dependencies on/for this PR:
* `master`
  * **PR #33409** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/33409?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 
    * **PR #33607** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/33607?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a>  👈
      * **PR #33608** <a href="https://app.graphite.dev/github/pr/airbytehq/airbyte/33608?utm_source=stack-comment-icon" target="_blank"><img src="https://static.graphite.dev/graphite-32x32-black.png" alt="Graphite" width="10px" height="10px"/></a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/airbytehq/airbyte/33607?utm_source=stack-comment).

aaronsteers commented on airbyte-lib/airbyte_lib/caches/memory.py:
If we are fine to only have local file caches and not in-memory caches, then we can delete this implementation.

aaronsteers commented on airbyte-lib/airbyte_lib/caches/snowflake.py:
Doesn't work yet. Probably is not a heavy lift but I didn't have a test harness set up for it, and did Postgres first because of the ease of testing locally and not needing creds.

aaronsteers commented on airbyte-lib/airbyte_lib/file_writers/base.py:
```diff
@@ -0,0 +1,70 @@
+"""Define abstract base class for Caches, which write and read from durable storage."""
```
```suggestion
"""Define abstract base class for File Writers, which write and read from file storage."""
```

aaronsteers commented on airbyte-lib/airbyte_lib/config.py:
```diff
@@ -0,0 +1,10 @@
+"""Define abstract base class for Caches, which write and read from durable storage."""
```
```suggestion
"""Define base Config interface, used by Caches and also File Writers (Processors)."""
```

aaronsteers commented on airbyte-lib/airbyte_lib/datasets.py:
Renamed file to be plural, to match other file name conventions, and allow for expansion if we make this a submodule layer.

aaronsteers commented on airbyte-lib/airbyte_lib/processors.py:
```diff
@@ -0,0 +1,266 @@
+"""Define abstract base class for Caches, which write and read from durable storage."""
```
```suggestion
"""Define abstract base class for Processors, including Caches and File writers.

Processors can all take input from STDIN or a stream of Airbyte messages.

Caches will pass their input to the File Writer. They share a common base class so certain abstractions like "write" and "finalize" can be handled in either layer, or both.
```

aaronsteers commented on airbyte-lib/airbyte_lib/sync_result.py:
Moved/renamed to "datasets.py" and "sync_results.py".

aaronsteers commented on airbyte-lib/airbyte_lib/types.py:
```diff
@@ -0,0 +1,80 @@
+"""Type conversion methods for SQL Caches."""
```
This is a generic implementation and will need work as we test in diverse scenarios. The abstractions, I think, are solid though - so ideally we will just need to incrementally add handling when we see failures on edge cases.

Still TBD if we will need to override this for different SQL dialects, but so far, it seems to work nicely generic. If we do need to override, the Cache class already has a means to take an overriden subclass of this main implementation, and those overriden implementations could have special handling if we find that some cases can't be implemented generically.

aaronsteers commented on airbyte-lib/examples/run_faker.py:
I'm hopeful to get source-faker working soon so we can test with large datasets and a more diverse set of data types.

aaronsteers replied:
Done. In separate PR.

aaronsteers commented on airbyte-lib/examples/run_test_source.py:
```diff
@@ -14,7 +14,7 @@
 os.environ["AIRBYTE_LOCAL_REGISTRY"] = "./tests/integration_tests/fixtures/registry.json"
 
 source = ab.get_connector("source-test", config={"apiKey": "test"})
-cache = ab.get_in_memory_cache()
+cache = ab.new_local_cache(source_catalog=source.configured_catalog)
```
Caches need a source catalog now, FYI. This informs data type decisions.

aaronsteers replied:
Update: this is now captured in the read() method so it is not needed in constructors and factory functions.

aaronsteers commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -0,0 +1,639 @@
+    """Same as a regular config except it exposes the 'get_sql_alchemy_url()' method."""
+
+    dedupe_mode = RecordDedupeMode.APPEND
+    schema_name: str = "airbyte_raw"
+    table_prefix: str = ""
+    table_suffix: str = ""
```
These mirror the config options on destinations.

flash1293 commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -0,0 +1,591 @@
+class SQLCacheConfigBase(CacheConfigBase):
+    """Same as a regular config except it exposes the 'get_sql_alchemy_url()' method."""
+
+    dedupe_mode = RecordDedupeMode.APPEND
```
Worth a discussion, but I expected the default mode to be overwrite which means it will remove the old table and recreate it. Right now with append it will fail on schema changes with no easy way to fix except for manually deleting the table from the destination. What do you think about introducing overwrite here which deletes the target table before loading new data?

Actually, append doesn't feel that important to me in general, especially since we don't have incremental syncs yet.

aaronsteers replied:
I've updated this significantly. Now, if primary keys exist, a merge insert will be attempted. If primary keys do not exist, then a table swap will be attempted.

We can continue to iterate as we improve our tests, in order to handle all the possible cases.

flash1293 commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -0,0 +1,591 @@
+            final_table_name = self._ensure_final_table_exists(stream_name, create_if_missing=True)
+
+            temp_table_name = self._write_files_to_new_table(files, stream_name, max_batch_id)
+            self._write_temp_table_to_final_table(stream_name, temp_table_name, final_table_name)
```
if the schema doesn't match the existing data in the cache, this step will fail, and the `_drop_temp_table` call won't get executed anymore, causing the temporary table to stick around.

Not necessarily a big issue, but I think we should add some logic here to try and clean up any WIP stuff in case something goes wrong.
<img width="270" alt="Screenshot 2024-01-03 at 12 15 48" src="https://github.com/airbytehq/airbyte/assets/1508364/ec2bb940-f8ec-4f32-a3f9-52715bab6346">

flash1293 commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -0,0 +1,591 @@
+    ) -> None:
+        """Return a new (unique) temporary table name."""
+        schema_name = self.config.schema_name
+        if schema_name in self._get_schemas_list():
```
This makes sure the table exists, but it won't make sure the column types are set up correctly. Should we add this here?

aaronsteers replied:
Yes, but probably out of scope for this iteration. Adding as a TODO.

aaronsteers replied:
I've added a new `_ensure_compatible_table_schema` method to handle this. For now, it just checks all columns exist and (if needed), the primary keys are defined.

flash1293 commented on airbyte-lib/airbyte_lib/file_writers/base.py:
```diff
@@ -0,0 +1,70 @@
+
+
+# The batch handle for file writers is a list of Path objects.
+@dataclass
```
There is no logic to clean up the files for the batches after the sync is finished - IMHO we should make sure that after a successful sync, no temporary stuff is left behind

aaronsteers replied:
Makes sense! 👍

aaronsteers replied:
Added cleanup steps for deleting files.

flash1293 commented on airbyte-lib/airbyte_lib/caches/duckdb.py:
```diff
@@ -0,0 +1,70 @@
+
+
+class DuckDBCache(DuckDBCacheBase):
+    """A DuckDB implementation of the cache.
```
We should add a TODO here to overwrite the naive `_write_files_to_new_table` implementation using the built-in duckdb functionality to read from a parquet file

aaronsteers replied:
Agreed. 👍

aaronsteers replied:
Done.

flash1293 commented on airbyte-lib/airbyte_lib/caches/postgres.py:
```diff
@@ -0,0 +1,47 @@
+        return self.database
+
+
+class PostgresCache(SQLCacheBase):
```
Will postgres keep using the naive loading via SQL or should we add a todo here to directly load the parquet files? Via a quick google search it doesn't look like there is an obvious way to load parquet into postgres in a performant way.

aaronsteers replied:
That's right. Postgres is the one case I noted in our initial set of in-scope DBs that does _not_ have native Parquet handling, at least not without plugins/extensions.

aaronsteers replied:
Added "TODO" comment with context here.

flash1293 commented on airbyte-lib/examples/run_spacex.py:
```diff
@@ -11,16 +11,20 @@
-print(islice(source.read_stream("capsules"), 10))
+print(islice(source.get_stream_records("capsules"), 10))
 
 for name, records in result.cache.streams.items():
```
It seems like those are outdated - `result` doesn't have a direct `cache` member anymore

aaronsteers replied:
I didn't retest these `examples/run_*.py` scripts - I only made small updates to match naming.

@flash1293 - Should we update these so they still work as expected? Are they helpful?

aaronsteers replied:
Done. Added back "cache" property for ReadResult (previously SyncResult).

flash1293 commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -0,0 +1,591 @@
+        return self.sql_alchemy_url
+
+
+class SQLCacheBase(RecordProcessor):
```
Per the spec, the cache should advertise all streams it knows about via `streams.items()`:
```
for stream_name, stream in cache.streams.items():
```

Seems like this part is missing (this is related to the comment in the examples which don't work anymore)

aaronsteers replied:
I've added `streams` for caches, although for now it is not implemented.

flash1293 reviewed (commented):
Added a bunch of comments, the biggest ones are:
* Some questions about the user-facing functions and classes
* Behavior on overwrite
* Cache access functionality doesn't seem to follow the spec

I'm OK with deferring some of them, but then we should add TODOs in the code (or create Github issues) to make sure they are tracked

aaronsteers commented on airbyte-lib/pyproject.toml:
```diff
@@ -1,29 +1,146 @@
-requires = ["poetry-core"]
-build-backend = "poetry.core.masonry.api"
+[tool.pytest.ini_options]
+# addopts = "--mypy"  # DEBUG: This blocks test discovery and execution, so I'm disabling it for now
```
Sure thing.

In the VS Code IDE, test discovery was failing when there were mypy issues, causing the test interface to essentially disappear - and you have to go digging in a different output window to find the mypy issue. I can debug a bit more, but that's the symptom I was dealing with. Since it was slowing me down, I disabled for my work. I can try enabling again and see if there are ways around it.

flash1293 replied:
The thing that's important to me is that CI is checking mypy for every PR - if we can find another way to ensure that would be fine for me as well.

Currently, airbyte-lib is tested via airbyte-ci which just executes pytest, that's what makes this integration attractive.

aaronsteers commented on airbyte-lib/tests/conftest.py:
```diff
@@ -0,0 +1,66 @@
+
+
+@pytest.fixture(scope="session", autouse=True)
+def remove_postgres_container():
```
Yes, that's exactly correct. Whenever you are debugging tests and you directly kill the debug process, the cleanup will (may?) be missed. This was added to make sure new sessions could be created without manually cleaning up old remnants.

aaronsteers commented:
@flash1293 - Thanks for this detailed feedback and for the extra pair of eyes. I'll update per these comments and ping again when it's ready for another review.

aaronsteers commented on airbyte-lib/tests/integration_tests/test_integration.py:
```diff
@@ -58,40 +72,92 @@ def test_check_fail():
+
     source = ab.get_connector("source-test", config={"apiKey": "test"})
-    cache = ab.get_in_memory_cache()
+    cache = ab.new_local_cache(source_catalog=source.configured_catalog)
```
I imagine there are two use cases, with an evolution from the first to the second over time.

At first, the user/developer will just want a local cache and not care how it is implemented. Later, the user may want to be more particular about where the data is stored and specifically how the cache is configured. The helper factory functions `new_local_cache()` and `get_default_cache()` are designed for the first use case, with "new_local_cache" biasing towards isolation of this cache from prior caches and "get_default_cache" biasing towards recycled/combined outputs because of the re-use of the same backend DB. In these cases, the user doesn't care what the backend is or how files are written, just that their data is stored somewhere and other behaviors are predictable. (DuckDB is intentionally not mentioned in the function name, and we could change implementation across versions, likely without negatively affecting the user.)

At some point, I expect the user to outgrow these functions and migrate to SnowflakeCache, PostgresCache, or similar - which have constructors designed to help with the more complex configuration required for those backends.

It feels weird to me that we might have a separate helper function for each type, like get_snowflake_cache, get_duckdb_cache, etc. Perhaps we could pass a factor function the CacheConfig object and let it decide the class based on the config object's class (or similar), but the user still needs to know a set of class names (and import them) in order to get type hinting, auto-complete, etc.

Wdyt of this approach and these relative benefits of class-based vs factory-based construction?

flash1293 replied:
Thanks for explaining, it's just something I stumbled over when looking at the code.

Your explanation makes sense to me, I think we can keep it like this, but if these classes are meant to be imported by the user IMHO we should clean up a little the exports of the `caches` module and the classes itself, as they expose a lot of stuff at the moment.

aaronsteers commented:
Reverted to draft while I finalize discussed updates. Will revert to "ready" once updates are complete - hopefully by EOD tomorrow.

aaronsteers commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -0,0 +1,591 @@
+    from sqlalchemy.engine.reflection import Inspector
+
+
+DEBUG_MODE = True
```
Updated to be 'False' by default. Consumers can still import it and manually set to 'True' in order to get more verbose logging.

aaronsteers commented on airbyte-lib/airbyte_lib/caches/memory.py:
```diff
@@ -0,0 +1,87 @@
+
+"""A DuckDB in-memory implementation of the cache.
+
+TODO: FIXME: In-memory caches don't work yet. Delete or fix.
```
As discussed offline, the only limitation really with the in-memory implementation is managing the connection object lifecycle. For now, local files are probably sufficient, and we can add in-memory option as a feature when time allows.

aaronsteers commented on airbyte-lib/examples/run_test_source.py:
```diff
@@ -14,7 +14,7 @@
 os.environ["AIRBYTE_LOCAL_REGISTRY"] = "./tests/integration_tests/fixtures/registry.json"
 
 source = ab.get_connector("source-test", config={"apiKey": "test"})
-cache = ab.get_in_memory_cache()
```
Good points, and I agree. I've updated `read()` (previously `read_all()`) to handoff the source name and configured catalog within that method via a new `register_catalog()` method.

aaronsteers commented on airbyte-lib/airbyte_lib/file_writers/__init__.py:
Done. 👍

aaronsteers commented on airbyte-lib/airbyte_lib/factories.py:
```diff
@@ -3,14 +3,47 @@
+
+
+def new_local_cache(
+    cache_name: str | None = None,
```
This behavior is intentional, specifically to avoid side effects of previous runs. When preferring accessibility and re-use, the user should give a name or use the similar factory function `get_default_cache()` which performs as you describe.

The proposed cleanup method (if/when needed) would be to simply delete the parent directory if remnants get out of hand. We use a cache directory which is a subdirectory to `cwd` in order to make this easy to explore and/or cleanup as needed.

aaronsteers commented:
@flash1293 - I've implemented all changes noted or discussed in review feedback. I've also implemented several renames that we discussed yesterday and today (although probably not all).

I've also added new pytest tests for linting - one for `ruff format` (passing), one for `ruff check` (not passing, xfail), and one for `mypy` (passing). In a future PR, we can burn down the `ruff check` issues.

aaronsteers commented on airbyte-lib/airbyte_lib/factories/_cache_factories.py:
Cache factories and connector factories were split into separate files in order to avoid a circular import. Sources need the cache factory function, and previously this same file was importing from source.py, creating a circular dependency. Resolved by splitting the files, so now source.py only needs to depend on the cache factory file, and there's no circular dependency back to source.py in that file.

aaronsteers commented on airbyte-lib/airbyte_lib/registry.py:
```diff
@@ -24,10 +23,12 @@ class ConnectorMetadata:
     global _cache
     if os.environ.get("AIRBYTE_LOCAL_REGISTRY"):
-        with open(str(os.environ.get("AIRBYTE_LOCAL_REGISTRY")), "r") as f:
+        with open(str(os.environ.get("AIRBYTE_LOCAL_REGISTRY"))) as f:
```
This was an auto-fix, so presumably the new is equivalent with the previous.

aaronsteers commented:
FYI - the formatting CI check is failing, but only because it is not capturing the preferences within this subdirectory, especially line length.

<img width="539" alt="image" src="https://github.com/airbytehq/airbyte/assets/18150651/160d97fb-cb6c-4c1a-ae6b-b91dd3cd4d0a">

flash1293 commented:
> FYI - the formatting CI check is failing, but only because it is not capturing the preferences within this subdirectory, especially line length.

I looked into this a bit and it's because we do a single call from the top level here: https://github.com/airbytehq/airbyte/blob/f90d8dbccb768e62cbddce178d6ec3af8d4ac8ea/airbyte-ci/connectors/pipelines/pipelines/airbyte_ci/format/configuration.py#L51

Apparently it is possible to respect nested config files as described here via the `--resolve-all-configs` flag.

https://pycqa.github.io/isort/docs/configuration/config_files.html#supporting-multiple-config-files-in-single-isort-run

I tried to hook that in, but it didn't recognize it (might be a versioning issue).

IMHO we should just use the regular top level config for formatting and not bother with separate settings just for airbyte-lib - it doesn't feel worth it to me. Also ccing @alafanechere here, maybe this came up already in some context.

What do you think @aaronsteers ?

flash1293 commented on airbyte-lib/airbyte_lib/factories/_cache_factories.py:
```diff
@@ -0,0 +1,33 @@
+
+
+def new_local_cache(
+    cache_name: str | None = None,
```
should we just make this the full path so users can store their duckdb file wherever?

aaronsteers replied:
My thought was that this would just be used for convenience - where you care only about controlling "reuse" and "side effects" between invocations over multiple executions, but you don't care as much about where/how the data is stored. 

That's the main reason duckdb isn't mentioned here - trying to remove friction. The "name" establishes a unique ID for the cache for purposes of having confidence that you are writing to a new or existing cache - but doesn't expose more internals than that. (For instance, the user doesn't have to worry about what file extension to use or what directories are writeable and git ignored.)

For full control, I think users would need to directly create a DuckDBCache object instead of using this factory function. I'm happy to revisit this discussion though in follow-ons. Probably helpful to think through if we want to use Pydantic for config classes of Snowflake, Postgres, BigQuery, etc. Then, how we treat the "full control" DuckDB cache can follow the same pattern.

aaronsteers replied:
I think just to give a balance and not be too magical, I'll add an optional arg to override the base directory. This, I think, gives a nice added layer of (optional) control while not requiring an understanding of DuckDB. And being optional will allow AirbyteLib to give a sensible default so the user doesn't need to go spelunking (for instance, especially in notebooks or other runtimes where they might not be familiar with the local filesystem).

aaronsteers replied:
Done. A new optional "root_dir" arg now allows the base directory to be overridden.

flash1293 commented on airbyte-lib/airbyte_lib/datasets/_map.py:
```diff
@@ -0,0 +1,23 @@
+from airbyte_lib.datasets._base import DatasetBase
+
+
+class DatasetMap(Mapping):
```
This doesn't seem to be used, is it meant to form the basis of ReadResult? (related to my other comment about overriding `items()` on the read result, I guess this would be the way to do it.

aaronsteers replied:
Yep. This is still only half baked, but that's the goal.

flash1293 commented on airbyte-lib/airbyte_lib/_util/airbyte.py:
super nit: could we rename this to protocol_utils or something? It being called `airbyte` confused me a bit at first.

aaronsteers replied:
Yeah. Totally! I thought the same.

aaronsteers replied:
Done. ✅

Now renamed to `protocol_util`

flash1293 reviewed (commented):
Thanks for the pass, left a few comments.

The biggest ones are the two bugs:
* Can't read from the iterator returned for a cached dataset
* Batch files are not deleted properly

plus some nits

flash1293 commented on airbyte-lib/airbyte_lib/types.py:
```diff
@@ -0,0 +1,88 @@
+    ) -> sqlalchemy.types.TypeEngine:
+        """Convert a value to a SQL type."""
+        airbyte_type, airbyte_subtype = _get_airbyte_type(json_schema_property_def)
+        json_schema_type = json_schema_property_def.get("type", None)
```
A big one that's not captured by this is dates - it feels like it would be worth including this as dates are very common data types and having them change types when migrating to hosted Airbyte would cause serious issues in downstream processing

aaronsteers replied:
Added unit tests and confirmed dates and datetimes are correctly covered. 👍

aaronsteers commented on airbyte-lib/tests/lint_tests/test_mypy.py:
```diff
@@ -0,0 +1,29 @@
+
+import pytest
+
+XFAIL = False  # Toggle to set if the test is expected to fail or not
```
Yes. Happy to remove here because MyPy is fine now. I'd like to leave the other Ruff one until we do the full cleanup.

aaronsteers commented on airbyte-lib/airbyte_lib/factories/__init__.py:
Yes, users can still import from the higher level models (top level or from "factories"). The lower level import only matters for our internal code, where we have to manage dependent DAG not being cyclical.

aaronsteers replied:
For now I removed imports from 'factories'. Just as a sidebar for us to dive into offline: I personally find that imports at multiple levels are really helpful. I appreciate that a lot of things can be imported from the top of the module, but also that there are submodules where I can browse a smaller list.

Still, we should of course discuss and be deliberate. For now, I don't know if / how much value is really added by having a 'factories' leaf for browsing/discovering these. Gone for now, but if `airbyte_lib` gets overly full, I think we can consider surfacing also in other tiers.

aaronsteers commented on airbyte-lib/airbyte_lib/results.py:
```diff
@@ -0,0 +1,21 @@
+    def __init__(self, processed_records: int, cache: SQLCacheBase) -> None:
+        self.processed_records = processed_records
+        self._cache = cache
+
```
Agreed this will make sense to add before final release. 👍

aaronsteers replied:
I started working on the in respect to caches (not fully done) and I imagine something similar for ReadResult as well.

aaronsteers commented on airbyte-lib/airbyte_lib/caches/memory.py:
```diff
@@ -0,0 +1,84 @@
+TODO: FIXME: In-memory caches don't work yet. Delete or fix.
+"""
+
+from __future__ import annotations
```
There are several places in the code where we have scaffolding but open bugs. When they are lower priority and not release blockers, I generally tried to put "fix or delete" type comments. That's the case here. I have like a 80-90% implementation for in-memory, but the "Singleton connection" requirement breaks a few things. 

Probably a time-boxed investment makes sense, if we have time, and if not, or if the feature takes more time than that, we'll delete this file and others related.

Does that make sense?

aaronsteers replied:
I went ahead and deleted the in-memory implementation. There are some remnants, specifically the 'singleton connection' implementation, which we can further refactor or remove in the future.

aaronsteers commented on airbyte-lib/airbyte_lib/caches/__init__.py:
```diff
@@ -0,0 +1,20 @@
+    "PostgresCache",
+    "PostgresCacheConfig",
+    "SQLCacheBase",
+    "SQLTypeConverter",
```
Good question. Originally I wanted it public because I'm still in the mode of "connectors" receiving contributions - and/or being BYO. But this is only needed (in theory) for overriding behaviors in different Caches, so I guess we can hide it because we don't expect users to build/customize/debug those cache implementations.

aaronsteers replied:
Update: Now removed from the imports surfaced in this "caches" module.

aaronsteers commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -0,0 +1,714 @@
+        stmt = table_ref.select()
+        with self.get_sql_connection() as conn:
+            curs: CursorResult[Any] = conn.execute(stmt)
+            return LazyDataset(curs)
```
Makes sense. Admittedly, I think this was a late addition, and I'm aware it is not yet where we want it to be. Same comment for iterating over the set of streams/datasets in general.

aaronsteers replied:
I reverted back to the previous implementation which simply "yields from" an iterator of dictionaries. I've placed a "TODO" block mentioning that we plan to replace these instances with a "LazyDataset" implementation.

aaronsteers commented on airbyte-lib/airbyte_lib/_processors.py:
```diff
@@ -0,0 +1,302 @@
+        self._finalized_state_messages[stream_name] += state_messages_to_finalize
+
+        for batch_id, batch_handle in batches_to_finalize.items():
+            self._cleanup_batch(stream_name, batch_id, batch_handle)
```
The file writer does have a cleanup method as of now, and it inherits from here. So when this runs in the file writer class, it should (in theory) correctly delete the files. I didn't have time to test but I at least wanted to get the implementation "hooks" in place.

Caches can also implement their own cleanup, if applicable, but they shouldn't have to cleanup files because the file writer class should handle that.

aaronsteers replied:
I realized the no-op implementation gave a false inference. Expanded docstrings here, to explain this better:

- https://github.com/airbytehq/airbyte/pull/33607/commits/1d6d079a1f75597ca0be0534d5215791cb954668

aaronsteers replied:
New integration test is added which confirms:

1. Files are written where expected.
2. Files are deleted afterwards.

I introduce "cleanup" as a boolean in the file writer config which makes it easy to test this - and also supports use cases where we prefer not to cleanup the files afterwards.

aaronsteers commented:
I set `airbyte-ci` to ignore `airbyte-lib` for both `black` and `isort`, due to project-level conflicts. The long-term path is to replace `black` and `isort` with `ruff`, as was proposed here:

- https://github.com/airbytehq/airbyte/pull/33251

Ruff is mono-repo aware, and will automatically use the closest-adjacent config. However, neither `black` nor `isort` is mono-repo aware.

In the case of `isort`, the main problem is in not being able to correctly identify first-party imports. (E.g., it cannot tell the difference between `airbyte_lib` and `airbyte_cdk`, and `airbyte_protocol` imports - and which is 'local' within each of these respective subprojects.) Ruff can do this automatically, so the path here is to just migrate to Ruff.

In the case of `black`, the main problem is that `140` is too wide as a default for this project. Especially when comparing diffs, and when `black`-style formatting auto-extends code to fully utilize the max width, this hurts readability and especially code review (where line width is doubled to review side-by-side).

aaronsteers commented on airbyte-lib/tests/conftest.py:
```diff
@@ -0,0 +1,64 @@
+        ports={"5432/tcp": PYTEST_POSTGRES_PORT},
+        detach=True,
+    )
+    # Wait for the database to start
```
I didn't really follow that comment and probably because it hasn't caused any issues, I've just left as-is. Can we move to a follow-on?

aaronsteers replied:
Now resolved.

flash1293 commented:
About formatting: Makes sense, your approach looks like a good path forward

flash1293 commented on airbyte-lib/airbyte_lib/_processors.py:
```diff
@@ -0,0 +1,309 @@
+    """Raised when an Airbyte message is invalid or cannot be parsed."""
+
+
+class RecordProcessor(abc.ABC):
```
Starting a new thread so it shows up properly in the GH UI (related to https://github.com/airbytehq/airbyte/pull/33607#discussion_r1447800745 )

Thanks for explaining a bit, I think I get it now - the weird part is the fact that both the Cache instance and the FileWriter instance are RecordProcessors, but the cache is managing the file writer. IMHO this makes the logic very difficult to follow because it's the same methods but they are called on different objects managing different state.

I think this comes from the fact that we wanted to provide file caches at some point as well, but it strikes me as too "clever" - even though it prevents redundancy, it makes it hard to follow to control flow.

We don't need to refactor on this PR because like you I would really like to get that one in, but IMHO we should make sure that the bug is fixed before we release.

I tried to make sense of the control flow once more and this is what happens:
* A batch is processed on the cache instance (`_process_batch`)
* It is delegating the actual writing of the batch to the underlying file writer (`_write_batch`), but returns the batch handle returned from that, so now the `_pending_batches` of the cache instance are holding the file writer batch handles pointing to the files
* Afterwards it's finalizing the batches by writing them to the actual duckdb (`_finalize_batches`), calling `_finalizing_batches` for cleanup, which does call `_cleanup_batch` with the file writer batch handles, but it calls it on the cache instance, not on the file writer instance, so it just executes the noop and leaves the files alone

You wrote this:
> The file writer does have a cleanup method as of now, and it inherits from here. So when this runs in the file writer class, it should (in theory) correctly delete the files. I didn't have time to test but I at least wanted to get the implementation "hooks" in place.

The issue is that `_cleanup_batch` on the file writer is never called with the batch handle, as the cache manages it and it doesn't delegate the call.

The easiest way to fix it now is what I wrote in the other comment - just like `_write_batch` on the cache delegates to the underlying file writer, `_cleanup_batch` needs to as well:
```python
    @overrides
    def _write_batch(
        self,
        stream_name: str,
        batch_id: str,
        record_batch: pa.Table | pa.RecordBatch,
    ) -> FileWriterBatchHandle:
        """
        Process a record batch.

        Return the path to the cache file.
        """
        return self.file_writer.write_batch(stream_name, batch_id, record_batch)

    # this is the new part
    def _cleanup_batch(
        self,
        stream_name: str,
        batch_id: str,
        batch_handle: BatchHandle,
    ) -> None:
        self.file_writer._cleanup_batch(stream_name, batch_id, batch_handle)
```

because the file writer is not "remembering" the patches it worked on, it's just getting a `_write_batch` call and processes it, so the cache needs to make sure it's called.

The cache might do more cleanup stuff per batch at some point, but it needs to make sure it's also done for the file write instance it manages.

As mentioned above, I think we should revisit this architecture, but doesn't have to happen on this PR as long as we fix the bug.

aaronsteers replied:
Regarding the design, I'm open to other methods but I think there are some benefits to this model.

I'll fix the plumbing for the batch handing.

What I didn't have time for yesterday, but I can do today, is just to add an integration test that asserts files are deleted as expected.

flash1293 replied:
Sounds good, +1 for an integration test

aaronsteers replied:
Done. 👍 

Tests added and passing.

flash1293 commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -0,0 +1,717 @@
+                # TODO: Add back metadata columns (this breaks tests)
+                # dataframe["_airbyte_loaded_at"] = str(pd.Timestamp.now())
+
+                dataframe.to_sql(
```
I tested with a source that has date columns and the final resulting table has a varchar again.
I think it's caused by this logic - the pandas dataframe has the date column as a string and via the `if_exists="replace"`, it won't honor the real table type and fall back to whatever types the dataframe provides which won't have gone through our sql type converter logic.

So in summary, it will create the table with the right types via `_create_table_for_loading`, then throw away these types again during the pandas loading step.

When switching `if_exists` to `append`, it works as expected (as the table is empty before this `to_sql` call)

aaronsteers replied:
Thanks for the clarification. The pandas "to_sql" was a bit of a hack, written before we had any catalog or type information.

aaronsteers replied:
Found and fixed an error where pandas defaulted to "replace" instead of "append". Also was able to plug in our detected datatypes into the Pandas "to_sql" implementation - although in practice this should be a no-op because now the previous table creation method will not be replaced. I considered just letting Pandas write the table, but Pandas doesn't seem to support Primary Keys, which we'll probably want eventually.

aaronsteers replied:
I think in a future iteration, we can add more advanced tests to ensure all data types and columns match exactly, but for that to be worthwhile we also need more complex test datasets, including datasets with diverse types and PK configurations. I think this is sufficient for now, since we're just trying to meet the functional spec of writing the records and being able to read them back.

Also, this PR is huge and I don't want to add more scope.

flash1293 reviewed (commented):
Left some more comments, biggest issues:
* batch parquet files are not deleted, left a comment clarifying what I meant
* proper column types are not kept

aaronsteers commented on airbyte-lib/airbyte_lib/factories/__init__.py:
I did that first and then reverted because I still think the factories-level import could be helpful as a bucket. For now, it's just empty to avoid confusion on the "right" way to import. But yes, I think swapping where the underscores are could make sense if we decide this is not needed at all.

aaronsteers replied:
Updated. The "factories" module is renamed to "_factories", and it's submodules have the "_" removed from their names..

aaronsteers commented on airbyte-lib/airbyte_lib/caches/base.py:
```diff
@@ -0,0 +1,717 @@
+        stmt = table_ref.select()
+        with self.get_sql_connection() as conn:
+            for row in conn.execute(stmt):
+                yield row.__dict__  # TODO: Refactor to return a LazyDataset here.
```
Yeesh, makes sense. I'll add that test. 👍

aaronsteers replied:
Added "test_get_records_result_as_list" and "test_read_result_as_list".

Also improved the `Dataset` implementations, which now inherit from the "Iterator" class so they can properly function as `Iterators`/`Iterables`.

These are both passing now:

```
def test_read_result_as_list(expected_test_stream_data: dict[str, list[dict[str, str | int]]]):
    source = ab.get_connector("source-test", config={"apiKey": "test"})
    cache = ab.new_local_cache()

    result: ReadResult = source.read(cache)
    stream_1_list = list(result["stream1"])
    stream_2_list = list(result["stream2"])
    assert stream_1_list == expected_test_stream_data["stream1"]
    assert stream_2_list == expected_test_stream_data["stream2"]


def test_get_records_result_as_list(expected_test_stream_data: dict[str, list[dict[str, str | int]]]):
    source = ab.get_connector("source-test", config={"apiKey": "test"})
    cache = ab.new_local_cache()

    stream_1_list = list(source.get_records("stream1"))
    stream_2_list = list(source.get_records("stream2"))
    assert stream_1_list == expected_test_stream_data["stream1"]
    assert stream_2_list == expected_test_stream_data["stream2"]
```

aaronsteers commented:
> Left some more comments, biggest issues:
> 
> * batch parquet files are not deleted, left a comment clarifying what I meant
> * proper column types are not kept

@flash1293 - These issues are now resolved. Can you give a final review? And as a next step, we can make a list of issues to address in subsequent iterations. I'll do this from my side also, but I want to prioritize landing this and resolving the already mentioned outstanding issues first.

flash1293 commented on airbyte-lib/airbyte_lib/_factories/connector_factories.py:
```diff
@@ -3,24 +3,19 @@
     use_local_install: bool = False,
-    install_if_missing: bool = True,
-):
+    install_if_missing: bool = False,
```
Seems like this was a merge gone bad - could you revert that to `True` as default again?

aaronsteers replied:
You are right. I think this got clobbered when resolving conflicts. Fixed now.

flash1293 commented on airbyte-lib/airbyte_lib/_factories/cache_factories.py:
```diff
@@ -0,0 +1,55 @@
+            raise ValueError(f"Cache name '{cache_name}' can only contain alphanumeric characters")
+
+    cache_name = cache_name or str(ulid.ULID())
+    cache_dir = cache_dir or Path("./.cache/{cache_name}")
```
This is missing an `f"`, it's literally calling the directory `{cache_name}`.

aaronsteers replied:
Good catch! Thanks!

flash1293 commented on airbyte-lib/airbyte_lib/_factories/cache_factories.py:
```diff
@@ -0,0 +1,55 @@
+        cache_dir = Path(cache_dir)
+
+    config = DuckDBCacheConfig(
+        db_path=cache_dir / f"db_{cache_name}.duckdb",
```
This will produce files like `.cache/01HKYK4KDF0653863VQ79MRH43/db_01HKYK4KDF0653863VQ79MRH43.duckdb`

The repeated ulid is a little weird, but not a big deal - no need to change, just wanted to mention I stumbled over it a bit.

aaronsteers replied:
Yeah. I think this is necessary for now, just for isolation of each cache iteration. If they give a name, or if they use the default cache, it'll be a shorter name.

Maybe in a future pass, we can shorten the string to the first 10 chars or so, or some similar decision.

One factor here is that DuckDB usees the file name as the DB name. So, having the name repeated is pretty much required if we want DB name to match cache name. And the dedicated subfolder is because we want to keep all cache artifacts next to each other.

There are probably other good ways to organize these, but for now I think it works for what we need. Happy to discuss a next iteration that changes things up.

flash1293 commented on airbyte-lib/airbyte_lib/_factories/cache_factories.py:
```diff
@@ -0,0 +1,55 @@
+        if " " in cache_name:
+            raise ValueError(f"Cache name '{cache_name}' cannot contain spaces")
+
+        if not cache_name.isalnum():
```
should we allow at least underscores in here? Seems pretty common

aaronsteers replied:
Done.

flash1293 reviewed (approved):
Left some last comments that should be addressed, but other than that this LGTM - great work putting this together, this is great base for the project!

All follow-up things seem to be captured in our doc and/or TODO comments

aaronsteers commented on airbyte-lib/tests/conftest.py:
```diff
@@ -0,0 +1,66 @@
+        detach=True,
+    )
+    # Wait for the database to start
+    time.sleep(5)
```
This is improved now: (1) we explicitly pull the image if it is missing, so the pull won't count against the sleep time, and (2) we loop in a retry loop for 10 seconds, retrying once per second until we can confirm a connection.
